plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,102 +64,126 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
+                            }
+                            if (discreteValues.getValuesCount() == 1) {
+                                fullyPushedDomains.add(columnHandle);
+                                return format("%s = %s",
+                                        CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                                        discreteValues.getValues().stream().findFirst()
+                                                .orElseThrow());
+                            }
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                return null;
                             }
-                            String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
-                                    + Joiner.on(",").join(discreteValuesList.build()) + ")";
-                            return predicate;
+
+                            String inValues = discreteValues.getValues().stream()
+                                    .map(columnHandle.getCassandraType()::toCqlLiteral)
+                                    .collect(joining(","));
+                            fullyPushedDomains.add(columnHandle);
+                            return CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN (" + inValues + " )";
                         }
                         return null;
                     }, allOrNone -> null);
 
             if (predicateString == null) {
                 break;
             }
-            // IN restriction only on last clustering column for Cassandra version = 2.1
-            if (predicateString.contains(" IN (") && cassandraVersion.compareTo(VersionNumber.parse("2.2.0")) < 0 && currentClusteringColumn != (clusteringColumns.size() - 1)) {
-                break;
-            }
             clusteringColumnSql.add(predicateString);
-            domainsBuilder.put(columnHandle, domain);
             // Check for last clustering column should only be restricted by range condition
             if (predicateString.contains(">") || predicateString.contains("<")) {
                 break;
             }
-            currentClusteringColumn++;
+            allProcesedClusteringColumns++;
         }
         List<String> clusteringColumnPredicates = clusteringColumnSql.build();
 
-        return new ClusteringPushDownResult(domainsBuilder.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+        return new ClusteringPushDownResult(fullyPushedDomains.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+    }
+
+    /**
+     * IN restriction allowed only on last clustering column for Cassandra version <= 2.1
+     */
+    private static boolean isInStatementNotAllowed(List<CassandraColumnHandle> clusteringColumns, VersionNumber cassandraVersion, int currentlyProcessedClusteringColumn)
```

Comment:
`isIn...Allowed` (without "Not") would read better.

intellij's suggestion `Java | Data flow | Boolean method is always inverted` is not something i would follow unconditionally

also "InStatement" -> "InExpression"